wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -524,6 +528,113 @@ class KotlinGenerator private constructor(
+      val fieldName = "DEFAULT_" + nameAllocator[field].toUpperCase(Locale.US)
+      val fieldType = field.getClass().copy(nullable = false)
+      val fieldValue = fieldInitializer(type, field.type(), nameAllocator, field.default)
+      companionBuilder.addProperty(PropertySpec.builder(fieldName, fieldType)
+          .apply {
+            if (field.type().isScalar && field.type() != ProtoType.BYTES) {
+              addModifiers(KModifier.CONST)
+            } else {
+              jvmField()
+            }
+          }
+          .initializer(fieldValue)
+          .build())
+    }
+  }
+
+  private fun fieldInitializer(
+    messageType: MessageType,
+    protoType: ProtoType,
+    nameAllocator: NameAllocator,
+    value: Any?
+  ): CodeBlock {
+    val typeName = protoType.typeName
+    return when {
+      value is List<*> -> buildCodeBlock {
+        add("listOf(")
+        var first = true
+        for (element in value) {
+          if (!first) add(",")
+          first = false
+          add("\n⇥%L⇤", fieldInitializer(messageType, protoType, nameAllocator, element))
+        }
+        add(")")
+      }
+      value is Map<*, *> -> buildCodeBlock {
+        add("copy(")
+        var first = true
+        for ((entryKey, entryValue) in value) {
+          if (!first) add(",")
+          first = false
+          val protoMember = entryKey as ProtoMember
+          val field = schema.getField(protoMember)
+          val entryInitializer =
+              fieldInitializer(messageType, field.type(), nameAllocator, entryValue)
+          add("\n⇥%L = %L⇤", nameAllocator[field], entryInitializer)
+        }
+        add(")")
+      }
+      typeName == BOOLEAN -> CodeBlock.of("%L", value ?: false)
+      typeName == INT -> value.toIntFieldInitializer()
+      typeName == LONG -> value.toLongFieldInitializer()
+      typeName == FLOAT -> CodeBlock.of("%Lf", value?.toString() ?: 0f)
+      typeName == DOUBLE -> CodeBlock.of("%L", value?.toString() ?: 0)
+      typeName == STRING -> CodeBlock.of("%S", value ?: "")
+      typeName == ByteString::class.asTypeName() -> if (value == null) {
+        CodeBlock.of("%M", ByteString::class.asClassName().member("EMPTY"))
+      } else {
+        CodeBlock.of("%S.%M()!!", value.toString().encode(charset = Charsets.ISO_8859_1).base64(),
+            ByteString.Companion::class.asClassName().member("decodeBase64"))
+      }
+      protoType.isEnum && value != null -> CodeBlock.of("%T.%L", typeName, value)
+      else -> throw IllegalStateException("$protoType is not an allowed scalar type")
+    }
+  }
+
+  private fun Any?.toIntFieldInitializer(): CodeBlock = when (val int = valueToInt()) {
+    Int.MIN_VALUE -> CodeBlock.of("%T.MIN_VALUE", INT)
+    Int.MAX_VALUE -> CodeBlock.of("%T.MAX_VALUE", INT)
+    else -> CodeBlock.of("%L", int)
+  }
+
+  private fun Any?.valueToInt(): Int {
+    if (this == null) return 0
+    val string = toString()
```

Comment:
Doesn't work with `when (val string = toString())`?

Reply:
Should work!

Reply:
Wait, no, `when (val string = toString())` would require all branch conditions to resolve to String, whereas what I need is Boolean

Reply:
That's not how that works. It just scopes the local to live inside the when. Doesn't affect what you do in the branches.

Reply:
I must be missing something, cause when I follow Benoit's suggestion I get compilation errors on branch conditions saying `Incompatible types: Boolean and String`.

Reply:
I think you're both talking about something different.